### PR TITLE
LOOP-887 Keycloak updates

### PR DESCRIPTION
**Pull request by ps2**
https://tidepool.atlassian.net/browse/LOOP-887

Updates for OAuth2 based authorization.  Also brings in some translations from DIY (not used yet in Tidepool Loop), and a stubbed out RemoteCommand method required for Service protocol updates, also not used in Tidepool Loop.

### CODE REVIEW

File: TidepoolServiceKit/IdentifiableDatum.swift [hunk 1]
```diff
@@ -29,18 +29,18 @@ extension IdentifiableDatum {
...
-    func datumOrigin<T: TypedDatum>(for type: T.Type) -> TOrigin {
-        return datumOrigin(for: resolvedIdentifier(for: type))
-    }
+//    var datumOrigin: TOrigin {
+//        return datumOrigin(for: resolvedIdentifier)
+//    }
+//
+//    func datumOrigin<T: TypedDatum>(for type: T.Type) -> TOrigin {
+//        return datumOrigin(for: resolvedIdentifier(for: type))
+//    }
```
Comment (Camji55) [inline TidepoolServiceKit/IdentifiableDatum.swift]: Can this be removed?

File: TidepoolServiceKit/IdentifiableHKDatum.swift [hunk 2]
```diff
@@ -30,26 +30,26 @@ extension IdentifiableHKDatum {
...
-    func datumOrigin<T: TypedDatum>(for type: T.Type) -> TOrigin? {
-        return datumOrigin(for: resolvedIdentifier(for: type))
-    }
+//    var datumOrigin: TOrigin? {
+//        return datumOrigin(for: resolvedIdentifier)
+//    }
+//
+//    func datumOrigin<T: TypedDatum>(for type: T.Type) -> TOrigin? {
+//        return datumOrigin(for: resolvedIdentifier(for: type))
+//    }
```
Comment (Camji55) [inline TidepoolServiceKit/IdentifiableHKDatum.swift]: This too?

File: TidepoolServiceKit/TidepoolService.swift [hunk 3]
```diff
@@ -483,68 +567,66 @@ extension TidepoolService: RemoteDataService {
...
+            throw error
         }
 
+        let dataSetId = try await getCachedDataSetId()
+
         // TODO: This implementation is incorrect and will not record the correct history when data is updated. Currently waiting on
         // https://tidepool.atlassian.net/browse/BACK-815 for backend to support new API to capture full history of data changes.
         // This work will be covered in https://tidepool.atlassian.net/browse/LOOP-3943. For now just call createData with the
         // updated data as it will just overwrite the previous data with the updated data.
-        tapi.createData(data, dataSetId: dataSetId) { error in
```
Comment (Camji55) [inline TidepoolServiceKit/TidepoolService.swift:520]: Is the above comment still relevant after these changes?

File: TidepoolServiceKitUI/SettingsView.swift [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+//
+//  Created by Pete Schwamb on 1/27/23.
+//  Copyright © 2023 LoopKit Authors. All rights reserved.
+//
+
+import SwiftUI
+import TidepoolKit
+import TidepoolServiceKit
+
+@MainActor
```
Comment (Camji55) [inline TidepoolServiceKitUI/SettingsView.swift]: I think `View` will automatically be on the `MainActor` if it contains an `ObservableObject`

Comment (Camji55) [inline TidepoolServiceKitUI/SettingsView.swift]: I think `View` will automatically be on the `MainActor` if it contains an `ObservableObject`. Was there a specific need for this annotation?

File: TidepoolServiceKitUI/SettingsView.swift [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+            Button(LocalizedString("Delete Service", comment: "Button title to delete a service"), role: .destructive) {
+                service.deleteService()
+                dismiss?()
+            }
+        }
+        .task {
+            do {
+                environments = try await TEnvironment.fetchEnvironments()
+            } catch {
+
```
Comment (Camji55) [inline TidepoolServiceKitUI/SettingsView.swift:120]: Do we need to write to the error property here?
```suggestion
self.error = error
```

File: TidepoolServiceKitUI/SettingsView.swift [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        }
+
+        error = nil
+        isLoggingIn = true
+
+        Task {
+            do {
+                try await login?(selectedEnvironment)
+                isLoggingIn = false
+                //dismiss?()
```
Comment (Camji55) [inline TidepoolServiceKitUI/SettingsView.swift]: `dismiss` needed?

File: TidepoolServiceKitUI/TidepoolService+UI.swift [hunk 7]
```diff
@@ -7,19 +7,62 @@
...
+        navController.isNavigationBarHidden = true
+
+        Task {
+            let service = TidepoolService(hostIdentifier: pluginHost.hostIdentifier, hostVersion: pluginHost.hostVersion)
+
+            let settingsView = await SettingsView(service: service, login: { environment in
+                try await service.tapi.login(environment: environment, presenting: navController)
+                try await service.completeCreate()
+                await navController.notifyServiceCreatedAndOnboarded(service)
+                //await navController.notifyComplete()
```
Comment (Camji55) [inline TidepoolServiceKitUI/TidepoolService+UI.swift]: Needed?

File: TidepoolServiceKit/TidepoolService.swift (hunk 3 above)
Comment (ps2) [reply TidepoolServiceKit/TidepoolService.swift:520]: I believe this is still outstanding, but isn't a blocker, as the creation makes sure the latest data is uploaded, but we lose some history.

File: TidepoolServiceKitUI/SettingsView.swift (hunk 4 above)
Comment (ps2) [reply TidepoolServiceKitUI/SettingsView.swift]: Yes, you're correct. I removed this and purposefully made the login function (called from loginButtonTapped) return from a non-main thread, and it switched back over to main actor.

File: TidepoolServiceKitUI/SettingsView.swift (hunk 5 above)
Comment (ps2) [reply TidepoolServiceKitUI/SettingsView.swift:120]: No, the environments is a "secret" list, that shouldn't expose errors to the user on failure. But good question.

File: TidepoolServiceKitUI/SettingsView.swift (hunk 6 above)
Comment (ps2) [reply TidepoolServiceKitUI/SettingsView.swift]: No, will remove.

File: TidepoolServiceKitUI/TidepoolService+UI.swift (hunk 7 above)
Comment (ps2) [reply TidepoolServiceKitUI/TidepoolService+UI.swift]: No, will remove.

**Review by Camji55 — APPROVED** [review]

**Review by nhamming — APPROVED** [review]
LGTM

**Pull request merged**